### DISCUSSION TITLE
Rendering plotly offline

### DISCUSSION THREAD

**Discussion by TaiPhamD** [quarto-dev/quarto-cli#5420] [Q&A] [answered]
When I generate html output from a .qmd or .ipynb with plotly graphs. The generated html files depend on using Plotly JavaScript from Plotlys CDN. In an environment without internet access will prevent us from loading the graphs. Is there an easy way to tell quarto to embed plotlys JavaScript as a static asset ?

**Comment by cscheid** [comment]
Unfortunately not, and we can't control it (the Plotly jupyter notebook output loads content dynamically)

**Reply by TaiPhamD** [reply]
That's interesting because when I publish a jupyter notebook from Rstudio jupyter workbench using the "publish" icon. The published document on our rstudio connect doesn't appear to be using plotlys cdn and renders just fine .

But I have the plotly cdn unreachable issue of I manually upload the quarto generated html/asset files to Rstudio connect

**Reply by cscheid** [reply]
Huh. I might just be 100% wrong. Can you try adding 

```yaml
self-contained: true
embed-resources: true
```

on your YAML front matter?

**Reply by TaiPhamD** [reply]
Thank you for the suggestion. Adding these created a single HTML file instead of HTML + folder asset. The HTML file is self-contained but the self-contained code still references the plotly internet CDN URL for the plotly javascript file so I still have the same issue.

**Reply by cscheid** [reply]
In that case, I don't understand how you get a "offline plotly" in your jupyter notebook. We're not doing anything special on our side...

**Reply by TaiPhamD** [reply]
Based on using chrome tools to inspect the working published jupyter notebook . I believe the direct jupyter published on reconnect has code in there that just says 'require': 'plotly' and no specific URL. In this case I think it uses the local plotly JS that is part of RSConnect server platform. The Quarto has the require statement with plotly but uses a specific version from a CDN URL.

**Reply by TaiPhamD** [reply]
Here is the chrome inspect of the difference between juptyter direct publish to RSConnect vs quarto:

# Jupyter work bench direct publishing to RSConnect
### (It embedded the whole plotly javascript I deleted the middle to keep it short)
```
<script type="text/javascript">
        window.PlotlyConfig = {MathJaxConfig: 'local'};
        if (window.MathJax && window.MathJax.Hub && window.MathJax.Hub.Config) {window.MathJax.Hub.Config({SVG: {font: "STIX-Web"}});}
        if (typeof require !== 'undefined') {
        require.undef("plotly");
        define('plotly', function(require, exports, module) {
            /**
* plotly.js v2.18.2
* Copyright 2012-2023, Plotly, Inc.
* All rights reserved.
* Licensed under the MIT license
*/
/*! For license information please see plotly.min.js.LICENSE.txt */
!function(t,e){"object"==typeof exports&&"object"==typeof module?module.exports=e():"function"==typeof define&&define.amd?define([],e):"object"==typeof exports?exports.Plotly=e():t.Plotly=e()}(self,(function(){return function(){var t={98847:function(t,e,r){"use strict";var n=r(71828),i={"X,X div":'direction:ltr;font-family:"Open Sans",verdana,arial,sans-serif;margin:0;padding:0;',"X input,X button":'font-family:"Open Sans",verdana,arial,sans-serif;',"X input:focus,X button:focus":"outline:none;","X a":"text-decoration:none;","X a:hover":"text-decoration:none;","X .crisp":"shape-rendering:crispEdges;","X .user-select-none":"-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;-o-user-select:none;user-select:none;","X svg":"overflow:hidden;","X svg a":"fill:#447adb;","X svg a:hover":"fill:#3c6dc5;","X .main-svg":"position:absolute;top:0;left:0;pointer-events:none;","X .main-svg .draglayer":"pointer-events:all;","X .cursor-default":"cursor:default;","X .cursor-pointer":"cursor:pointer;","X .cursor-crosshair":"cursor:crosshair;","X .cursor-move":"cursor:move;","X .cursor-col-resize":"cursor:col-resize;","X .cursor-row-resize":"cursor:row-resize;","X …
 
[DELETED JAVASCRIPT CODE TO KEEP IT SHORT]
window}}(),r.o=function(t,e){return Object.prototype.hasOwnProperty.call(t,e)},r.r=function(t){"undefined"!=typeof Symbol&&Symbol.toStringTag&&Object.defineProperty(t,Symbol.toStringTag,{value:"Module"}),Object.defineProperty(t,"__esModule",{value:!0})},r(27909)}()}));
        });
        require(['plotly'], function(Plotly) {
            window._Plotly = Plotly;
        });
        }
</script>
```

# Quarto render to RSConnect

```
<script type="text/javascript">
        window.PlotlyConfig = {MathJaxConfig: 'local'};
        if (window.MathJax && window.MathJax.Hub && window.MathJax.Hub.Config) {window.MathJax.Hub.Config({SVG: {font: "STIX-Web"}});}
        if (typeof require !== 'undefined') {
        require.undef("plotly");
        requirejs.config({
            paths: {
                'plotly': ['https://cdn.plot.ly/plotly-2.18.2.min']
            }
        });
        require(['plotly'], function(Plotly) {
            window._Plotly = Plotly;
        });
        }
</script>
```

**Reply by cscheid** [reply]
> Jupyter work bench direct publishing to RSConnect

What I'm asking is: what steps did you take to get that .ipynb file as a result?

**Reply by TaiPhamD** [reply]
both publishing uses the same .ipynb file

jupyter direct publishing to RSConnect I just use the "publish" button as seen here:
<img width="1215" alt="Screenshot 2023-05-04 at 12 18 26 PM" src="https://user-images.githubusercontent.com/10516699/236307257-5220d9bc-d3c2-467f-b4a8-dfe661b6d807.png">

Then for Quarto render I just simply do:
quarto render [notebook].ipnb --execute

Then once I get the resultant html/asset file I use the R rsconnect:: api to push it to the same RSConnect server

**Reply by cscheid** [reply, marked as answer]
Ok, try adding this to the top of your file:

```{python}
import plotly.io as pio
pio.renderers.default = "notebook"
```

**Reply by TaiPhamD** [reply]
Thank you!! it's working now.
